neonvm/runner/containermgr.go

```diff
@@ -0,0 +1,364 @@
+package main
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"os/exec"
+	"strconv"
+	"sync/atomic"
+	"time"
+
+	"go.uber.org/zap"
+
+	vmv1 "github.com/neondatabase/autoscaling/neonvm/apis/neonvm/v1"
+	"github.com/neondatabase/autoscaling/pkg/api"
+)
+
+const (
+	runnerContainerName = "neonvm-runner"
+
+	retryInitEvery = 5 * time.Second
+
+	// cpuLimitOvercommitFactor sets the amount above the VM's spec.guest.cpus.use that we set the
+	// QEMU cgroup's CPU limit to. e.g. if cpuLimitOvercommitFactor = 3 and the VM is using 0.5
+	// CPUs, we set the cgroup to limit QEMU+VM to 1.5 CPUs.
+	//
+	// This exists because setting the cgroup exactly equal to the VM's CPU value is overly
+	// pessimistic, results in a lot of unused capacity on the host, and particularly impacts
+	// operations that parallelize between the VM and QEMU, like heavy disk access.
+	//
+	// See also: https://neondb.slack.com/archives/C03TN5G758R/p1693462680623239
+	cpuLimitOvercommitFactor = 4
+
+	// number of CPU shares per vCPU to use when updating a container
+	sharesPerCPU = 1024
+	cpuPeriod    = 100000
+)
+
+// mainContainerMgr is the 'main' function if neonvm-runner is called with the container-mgr bin path
+//
+// This typically runs as a secondary container alongside neonvm-runner, with privileged access to
+// conatinerd via crictl.
+func mainContainerMgr() {
+	logger := zap.Must(zap.NewProduction()).Named("neonvm-container-mgr")
+
+	selfPodUID, ok := os.LookupEnv("K8S_POD_UID")
+	if !ok {
+		logger.Fatal("environment variable K8S_POD_UID missing")
+	}
+	logger.Info("Got pod UID", zap.String("uid", selfPodUID))
+
+	containerRuntimeEndpoint, ok := os.LookupEnv("CRI_ENDPOINT")
+	if !ok {
+		logger.Fatal("environment variable CRI_ENDPOINT missing")
+	}
+	logger.Info("Got CRI endpoint", zap.String("endpoint", containerRuntimeEndpoint))
+
+	crictl := &crictlConfig{
+		endpoint: containerRuntimeEndpoint,
+	}
+
+	var httpPort int
+	flag.IntVar(&httpPort, "port", -1, "Port for the CPU http server")
+	flag.Parse()
+
+	var pods crictlPods
+	if err := crictl.run(logger, &pods, "pods", "-o", "json"); err != nil {
+		logger.Fatal("failed to run crictl command to get CRI ID for pod UID", zap.String("uid", selfPodUID), zap.Error(err))
+	}
+
+	// find pod with matching uid
+	var criPodID string
+	for _, p := range pods.Items {
+		if p.Metadata.UID == selfPodUID {
+			criPodID = p.ID
+			break
+		}
+	}
+	if criPodID == "" {
+		logger.Fatal("could not find CRI pod with matching UID", zap.String("uid", selfPodUID))
+	}
+
+	logger.Info("Got CRI ID for pod", zap.String("podID", criPodID))
+
+	var criRunnerContainerID string
+	for criRunnerContainerID == "" {
+		var containers crictlContainers
+		if err := crictl.run(logger, &containers, "ps", "-p", criPodID, "-o", "json"); err != nil {
```

Comment:
Let's add a comment describing what does this `crictl` command do. Another idea is providing a higher level abstraction on top of `run`, smth like `crictl.ListContainers(...)`. Also this helps us with type safety to some extent.

Reply:
I ended up going with this approach in d8f1b19b553fed039ecdaeb9f176e053c00d83af - resolving this comment in favor of the one above.